File: node/src/shared-strings.ts
import { closeSync, mkdtempSync, openSync, readSync, rmSync, writeSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import type { XlsxArchive } from "./zip.js";
import { parseXml } from "./xml.js";

export class SharedStringStore {
  readonly #directory: string;
  readonly #fd: number;
  readonly #offsets: number[] = [];
  #position = 0;

  private constructor() {
    this.#directory = mkdtempSync(join(tmpdir(), "minimarkdown-strings-"));
    this.#fd = openSync(join(this.#directory, "strings.bin"), "w+");
  }

  static async load(archive: XlsxArchive): Promise<SharedStringStore> {
    const store = new SharedStringStore();
    if (!archive.has("xl/sharedStrings.xml")) return store;
    let current: string | undefined;
    let inText = false;
    await archive.withEntry("xl/sharedStrings.xml", async (stream) => {
      await parseXml(stream, {
        open(tag) {
          if (tag.local === "si") current = "";
          else if (tag.local === "t" && current !== undefined) inText = true;
        },
        text(value) {
          if (inText && current !== undefined) current += value;
        },
        close(tag) {
          if (tag.local === "t") inText = false;
          else if (tag.local === "si") {
            store.#write(current ?? "");
            current = undefined;
          }
        },
      });
    });
    return store;
  }

  get(indexText: string): string {
    const index = Number(indexText);
    const offset = Number.isInteger(index) && index >= 0 ? this.#offsets[index] : undefined;
    if (offset === undefined) throw new Error("A shared string index is invalid.");
    const lengthBytes = Buffer.allocUnsafe(4);
    readSync(this.#fd, lengthBytes, 0, 4, offset);
    const value = Buffer.allocUnsafe(lengthBytes.readUInt32LE());
    readSync(this.#fd, value, 0, value.length, offset + 4);
    return value.toString("utf8");
  }

  close(): void {
    closeSync(this.#fd);
    rmSync(this.#directory, { recursive: true, force: true });
  }

  #write(value: string): void {
    const bytes = Buffer.from(value, "utf8");
    const length = Buffer.allocUnsafe(4);
    length.writeUInt32LE(bytes.length);
    this.#offsets.push(this.#position);
    writeSync(this.#fd, length, 0, length.length, this.#position);
    writeSync(this.#fd, bytes, 0, bytes.length, this.#position + 4);
    this.#position += 4 + bytes.length;
  }
}

File: rust/src/shared_strings.rs
use std::io::{self, BufReader, Read, Seek, SeekFrom, Write};

use quick_xml::Reader;
use quick_xml::events::Event;
use tempfile::NamedTempFile;
use zip::ZipArchive;

use crate::xml::{local_name, text_value};
use crate::{ConversionResult, invalid_data};

pub(crate) struct SharedStringStore {
    file: NamedTempFile,
    offsets: Vec<u64>,
}

impl SharedStringStore {
    pub(crate) fn load<R: Read + Seek>(archive: &mut ZipArchive<R>) -> ConversionResult<Self> {
        let mut store = Self {
            file: NamedTempFile::new()?,
            offsets: Vec::new(),
        };
        let Ok(entry) = archive.by_name("xl/sharedStrings.xml") else {
            return Ok(store);
        };
        let mut reader = Reader::from_reader(BufReader::new(entry));
        let mut buffer = Vec::new();
        let mut item: Option<String> = None;
        let mut in_text = false;

        loop {
            match reader.read_event_into(&mut buffer)? {
                Event::Start(element) => match local_name(element.name().as_ref()) {
                    b"si" => item = Some(String::new()),
                    b"t" if item.is_some() => in_text = true,
                    _ => {}
                },
                Event::Text(text) if in_text => {
                    item.as_mut().unwrap().push_str(&text_value(&text)?);
                }
                Event::CData(text) if in_text => {
                    item.as_mut().unwrap().push_str(&text.decode()?);
                }
                Event::End(element) => match local_name(element.name().as_ref()) {
                    b"t" => in_text = false,
                    b"si" => store.write(item.take().unwrap_or_default())?,
                    _ => {}
                },
                Event::Eof => break,
                _ => {}
            }
            buffer.clear();
        }
        Ok(store)
    }

    pub(crate) fn get(&mut self, index_text: &str) -> ConversionResult<String> {
        let index: usize = index_text
            .parse()
            .map_err(|_| invalid_data("A shared string index is invalid."))?;
        let offset = *self
            .offsets
            .get(index)
            .ok_or_else(|| invalid_data("A shared string index is invalid."))?;
        let file = self.file.as_file_mut();
        file.seek(SeekFrom::Start(offset))?;
        let mut length = [0_u8; 8];
        file.read_exact(&mut length)?;
        let length = u64::from_le_bytes(length);
        let length =
            usize::try_from(length).map_err(|_| invalid_data("A shared string is too large."))?;
        let mut bytes = vec![0_u8; length];
        file.read_exact(&mut bytes)?;
        String::from_utf8(bytes).map_err(|error| Box::new(error) as _)
    }

    fn write(&mut self, value: String) -> io::Result<()> {
        let file = self.file.as_file_mut();
        self.offsets.push(file.stream_position()?);
        file.write_all(&(value.len() as u64).to_le_bytes())?;
        file.write_all(value.as_bytes())
    }
}
